dbt/compilation.py

```diff
@@ -171,54 +186,96 @@ def do_ref(*args):
+                                                target_model.get('unique_id')))
 
-            # this creates a trivial cycle -- should this be a compiler error?
-            # we can still interpolate the name w/o making a self-cycle
-            if source_model == other_model_fqn:
-                pass
-            else:
-                linker.dependency(source_model, other_model_fqn)
+            model['depends_on'].append(target_model_id)
 
-            if other_model.is_ephemeral:
-                linker.inject_cte(model, other_model)
-                return other_model.cte_name
+            if target_model.get('config', {}) \
+                           .get('materialized') == 'ephemeral':
+
+                model['extra_cte_ids'].append(target_model_id)
+                return '__dbt__CTE__{}'.format(target_model.get('name'))
             else:
-                return '"{}"."{}"'.format(schema, other_model_name)
+                return '"{}"."{}"'.format(schema, target_model.get('name'))
 
         def wrapped_do_ref(*args):
             try:
                 return do_ref(*args)
             except RuntimeError as e:
-                root = os.path.relpath(
-                    model.root_dir,
-                    model.project['project-root']
-                )
-
-                filepath = os.path.join(root, model.rel_filepath)
-                logger.info("Compiler error in {}".format(filepath))
+                logger.info("Compiler error in {}".format(model.get('path')))
                 logger.info("Enabled models:")
-                for m in all_models:
-                    logger.info(" - {}".format(".".join(m.fqn)))
+                for n,m in all_models.items():
+                    logger.info(" - {}".format(".".join(m.get('fqn'))))
                 raise e
 
         return wrapped_do_ref
 
+    def get_compiler_context(self, linker, model, models,
+                            macro_generator=None):
+        context = self.project.context()
+
+        if macro_generator is not None:
+            for macro_data in macro_generator(context):
+                macro = macro_data["macro"]
+                macro_name = macro_data["name"]
+                project = macro_data["project"]
+
+                if context.get(project.get('name')) is None:
+                    context[project.get('name')] = {}
+
+                context.get(project.get('name'), {}) \
+                       .update({macro_name: macro})
+
+                if model.get('package_name') == project.get('name'):
+                    context.update({macro_name: macro})
+
+        adapter = get_adapter(self.project.run_environment())
+
+        # built-ins
+        context['ref'] = self.__ref(context, model, models)
+        context['config'] = self.__model_config(model, linker)
+        context['this'] = This(
+            context['env']['schema'],
+            (model.get('name') if dbt.flags.NON_DESTRUCTIVE
```

Comment:
this is ported from `model.tmp_name()`, right? it took me a minute to work out what was happening here... maybe we could add a comment like:

> for non-destructive runs, `this` should return the final name of the model, because we're truncating and inserting instead dropping and re-creating.